contracts/pallet-ibc/src/ics20_fee/mod.rs

```diff
@@ -0,0 +1,272 @@
+			relayer,
+		)
+	}
+
+	fn on_chan_open_ack(
+		&mut self,
+		ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		counterparty_version: &Version,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		self.inner
+			.on_chan_open_ack(ctx, output, port_id, channel_id, counterparty_version, relayer)
+	}
+
+	fn on_chan_open_confirm(
+		&mut self,
+		ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		self.inner.on_chan_open_confirm(ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_chan_close_init(
+		&mut self,
+		ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		self.inner.on_chan_close_init(ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_chan_close_confirm(
+		&mut self,
+		ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		self.inner.on_chan_close_confirm(ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_recv_packet(
+		&self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		packet: &Packet,
+		relayer: &Signer,
+	) -> Result<Acknowledgement, Ics04Error> {
+		// Module ModuleCallbackContext does not have the ics20 context as part of its trait bounds
+		// so we define a new context
+		let mut ctx = Context::<T>::default();
+		let packet_data =
+			serde_json::from_slice::<PacketData>(packet.data.as_slice()).map_err(|e| {
+				Ics04Error::implementation_specific(format!("Failed to decode packet data {:?}", e))
+			})?;
+		let percent = ServiceCharge::<T>::get().unwrap_or(T::ServiceCharge::get());
+		// Send full amount to receiver using the default ics20 logic
+		let ack = self.inner.on_recv_packet(&mut ctx, output, packet, relayer)?;
+		// We only take the fee charge if the acknowledgement is not an error
+		if ack.as_ref() ==
+			ibc::applications::transfer::acknowledgement::Acknowledgement::success()
+				.to_string()
+				.as_bytes()
+		{
+			// We have ensured that token amounts larger than the max value for a u128 are rejected
+			// in the ics20 on_recv_packet callback so we can multiply safely.
+			// Percent does Non-Overflowing multiplication so this is infallible
+			let fee = percent * packet_data.token.amount.as_u256().low_u128();
+			let receiver = <T as crate::Config>::AccountIdConversion::try_from(
+				packet_data.receiver,
+			)
+			.map_err(|_| {
+				Ics04Error::implementation_specific("Failed to receiver account".to_string())
+			})?;
+			let pallet_account = Pallet::<T>::account_id();
+			let mut prefixed_coin = if is_receiver_chain_source(
+				packet.source_port.clone(),
+				packet.source_channel,
+				&packet_data.token.denom,
+			) {
```

Comment:
you can simplify this entire block by outputting the channel and port and then calculating the `c` value only once

Reply:
We perform different functions on the c in the if blocks, we `remove_trace_prefix` in one and `add_trace_prefix` in the other.

Reply:
oop sorry